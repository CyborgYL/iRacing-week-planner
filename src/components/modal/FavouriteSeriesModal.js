import * as React from 'react';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';
import Modal from './Modal';
import Checkbox from '../Checkbox';
import series from '../../data/season.json';

import styles from '../../styles/main.module.scss';

const groupedSeries = series.reduce((grouped, single) => {
  if (single.catid === 1) {
    return {
      ...grouped,
      oval: [
        ...grouped.oval,
        single,
      ],
    };
  }
  if (single.catid === 3) {
    return {
      ...grouped,
      dirtOval: [
        ...grouped.dirtOval,
        single,
      ],
    };
  }
  if (single.catid === 4) {
    return {
      ...grouped,
      rx: [
        ...grouped.rx,
        single,
      ],
    };
  }
  return {
    ...grouped,
    road: [
      ...grouped.road,
      single,
    ],
  };
}, { oval: [], road: [], dirtOval: [], rx: [] });

class FavouriteSeriesModal extends React.Component {
  static propTypes = {
    onClose: PropTypes.func,
    isOpen: PropTypes.bool.isRequired,
    save: PropTypes.func,
    favouriteSeries: PropTypes.array,
    t: PropTypes.func.isRequired,
  }

  static defaultProps = {
    onClose: () => {},
    isOpen: false,
    save: () => {},
    favouriteSeries: [],
  }

  setCheckboxFavourite(seriesId, newValue) {
    const { favouriteSeries, save } = this.props;
    const newFavorites = [...favouriteSeries];
    const index = newFavorites.indexOf(seriesId);

    if (index === -1 && newValue) {
      newFavorites.push(seriesId);
    }
    if (index !== -1 && newValue === false) {
      newFavorites.splice(index, 1);
    }
    save(newFavorites);
  }

  render() {
    const { onClose, isOpen, t, favouriteSeries } = this.props;

    const renderCheckboxes = (cbSeries) => (
      <div className={styles['col-md-6']} key={cbSeries.seriesid}>
        <Checkbox
          id={`favourite-series-${cbSeries.seriesid}`}
          checked={favouriteSeries.indexOf(cbSeries.seriesid) !== -1}
          onChange={() => {
            this.setCheckboxFavourite(cbSeries.seriesid);
          }}
        >
          {t(cbSeries.seriesname)}
        </Checkbox>
      </div>
    );

    return (
      <Modal isOpen={isOpen} onClose={onClose} title={t('Choose favorite series')} doneAction={onClose}>
        <div className={styles['container-fluid']}>
          <h3>{t('Oval')}</h3>
          <div className={styles.row}>
            {groupedSeries.oval.map(renderCheckboxes)}
          </div>
          <h3>{t('Road')}</h3>
          <div className={styles.row}>
            {groupedSeries.road.map(renderCheckboxes)}
          </div>
          <h3>{t('Dirt Oval')}</h3>
          <div className={styles.row}>
            {groupedSeries.dirtOval.map(renderCheckboxes)}
          </div>
          <h3>{t('RX')}</h3>
          <div className={styles.row}>
            {groupedSeries.rx.map(renderCheckboxes)}
          </div>
        </div>
      </Modal>
    );
  }
}

export default withTranslation()(FavouriteSeriesModal);
