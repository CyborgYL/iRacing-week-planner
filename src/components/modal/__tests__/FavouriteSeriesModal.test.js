import { describe, test } from '@jest/globals';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import FavouriteSeriesModal from '../FavouriteSeriesModal';

jest.mock('firebase/auth');

describe('components/modal/FavouriteSeriesModal', () => {
  test('renders closed', async () => {
    const onClose = jest.fn(() => {});
    const component = renderer.create(
      <FavouriteSeriesModal isOpen={false} onClose={onClose} favouriteSeries={[]} save={() => {}} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders with checkbox saving', async () => {
    const onClose = jest.fn(() => {});
    const save = jest.fn(() => {});
    let component;
    act(() => {
      component = renderer.create(
        <FavouriteSeriesModal isOpen onClose={onClose} favouriteSeries={[139]} save={save} />,
      );
    });

    expect(component.toJSON()).toMatchSnapshot();

    await act(async () => {
      await component.root.findByProps({ id: 'favourite-series-139' }).props.onChange(false);
    });

    expect(save).toHaveBeenCalledWith([]);

    await act(async () => {
      await component.root.findByProps({ id: 'favourite-series-432' }).props.onChange(true);
    });

    expect(save).toHaveBeenCalledWith([139, 432]);

    await act(async () => {
      await component.root.findByProps({ 'aria-label': 'Close' }).props.onClick({
        preventDefault: () => {},
        stopPropagation: () => {},
      });
    });

    expect(onClose).toHaveBeenCalled();
  });
});
