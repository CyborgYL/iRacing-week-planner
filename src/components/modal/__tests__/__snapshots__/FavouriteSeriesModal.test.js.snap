// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/FavouriteSeriesModal renders closed 1`] = `null`;

exports[`components/modal/FavouriteSeriesModal renders with checkbox saving 1`] = `
<div
  className="ReactModal__Overlay"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content Modal__Bootstrap modal-dialog modal-lg"
    id="favouriteSeriesModal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Choose favorite series
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div
          className="container-fluid"
        >
          <h3>
            Oval
          </h3>
          <div
            className="row"
          >
            <div
              className="col-md-6"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="favourite-series-32"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="favourite-series-32"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  iRacing Advanced Legends Cup
                </label>
              </div>
            </div>
            <div
              className="col-md-6"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="favourite-series-167"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="favourite-series-167"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  iRacing ARCA Menards Series
                </label>
              </div>
            </div>
          </div>
          <h3>
            Road
          </h3>
          <div
            className="row"
          >
            <div
              className="col-md-6"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="favourite-series-139"
                >
                  <input
                    checked={true}
                    disabled={false}
                    id="favourite-series-139"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Fanatec Global Mazda MX-5 Cup
                </label>
              </div>
            </div>
            <div
              className="col-md-6"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="favourite-series-432"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="favourite-series-432"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Fanatec GT Challenge
                </label>
              </div>
            </div>
          </div>
          <h3>
            Dirt Oval
          </h3>
          <div
            className="row"
          >
            <div
              className="col-md-6"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="favourite-series-341"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="favourite-series-341"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Example Dirt Oval
                </label>
              </div>
            </div>
          </div>
          <h3>
            RX
          </h3>
          <div
            className="row"
          >
            <div
              className="col-md-6"
            >
              <div
                className="checkbox"
              >
                <label
                  htmlFor="favourite-series-323"
                >
                  <input
                    checked={false}
                    disabled={false}
                    id="favourite-series-323"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  Example Rallycross
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
